File: app/src/main/java/com/koko/www/androiddemo/activity/status/StateActivity.java
package com.koko.www.androiddemo.activity.status;

import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;

import androidx.fragment.app.Fragment;

import com.koko.www.androiddemo.R;

public class StateActivity extends AppCompatActivity {
    static final String SOME_VALUE = "int_value";
    static final String SOME_OTHER_VALUE = "string_value";
    public static String StateActivity_TAG = "StateActivity";
    private int someIntValue;
    private String someStringValue;
    private Fragment simpleFragment;
    private final String SIMPLE_FRAGMENT_TAG = "myfragmenttag";
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.e(StateActivity_TAG, "-----Activity/onCreate()------");
        setContentView(R.layout.activity_state);

        //确保在配置更改时不会不必要地重新创建片段。
        if (savedInstanceState != null) {
            simpleFragment = getSupportFragmentManager().findFragmentByTag(SIMPLE_FRAGMENT_TAG);
        } else if (simpleFragment == null) {
            // only create fragment if they haven't been instantiated already
            simpleFragment = new SimpleFragment();
            Log.e(StateActivity_TAG,"SimpleFragment被创建");
        }


    }

    @Override
    protected void onResume() {
        super.onResume();
        Log.e(StateActivity_TAG, "--------Activity/onResume()------");
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Log.e(StateActivity_TAG, "--------Activity/onDestroy()------");
    }

    //系统将在活动销毁之前调用该方法
    @Override
    protected void onSaveInstanceState(Bundle savedInstanceState) {
        Log.e(StateActivity_TAG, "----Activity/onSaveInstanceState()-----");
        // Save custom values into the bundle
        savedInstanceState.putInt(SOME_VALUE, 1);
        savedInstanceState.putString(SOME_OTHER_VALUE, "koko");
        // Always call the superclass so it can save the view hierarchy state
        super.onSaveInstanceState(savedInstanceState);

    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        Log.e(StateActivity_TAG, "----Activity/onRestoreInstanceState()-----");
        // Always call the superclass so it can restore the view hierarchy
        super.onRestoreInstanceState(savedInstanceState);
        // Restore state members from saved instance
        someIntValue = savedInstanceState.getInt(SOME_VALUE);
        someStringValue = savedInstanceState.getString(SOME_OTHER_VALUE);
        Log.e(StateActivity_TAG, "Activity/someIntValue=" + someIntValue + "//" + "someStringValue=" + someStringValue);
    }
}

File: app/src/main/java/com/koko/www/androiddemo/activity/status/StateActivity.kt
package com.koko.www.androiddemo.activity.status

import android.content.Context
import android.os.Bundle
import android.util.Log
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import com.koko.www.androiddemo.R

/**
 * 关于activity保存状态的demo
 * 比如旋转屏幕时
 *
 * 1.使用自带的Bundle保存部分状态
 * 2.使用SharedPreferences保存键值对来保存状态
 */
class StateActivity : AppCompatActivity() {
    private lateinit var editText: EditText
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_state)
        editText = findViewById(R.id.editText)

        //任意context都可以获取的
//        val sharedPref = this.getSharedPreferences(
//                getString(R.string.preference_file_key), Context.MODE_PRIVATE)

        //当前activity获取偏好
        val sharedPref1 = this.getPreferences(Context.MODE_PRIVATE) ?: return
        val defaultValue = 25
        val highScore = sharedPref1.getInt(getString(R.string.activity_pref_key), defaultValue)
        Toast.makeText(this, ""+highScore, Toast.LENGTH_SHORT).show()
    }


    //系统将在活动销毁之前调用该方法
    override fun onSaveInstanceState(savedInstanceState: Bundle) {
        savedInstanceState.putString("key", editText.text.toString())
        super.onSaveInstanceState(savedInstanceState)
    }

    //恢复状态时调用
    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        Toast.makeText(this, savedInstanceState.getString("key"), Toast.LENGTH_SHORT).show()
    }

    override fun onDestroy() {
        //当前activity获取的
        val sharedPref = getPreferences(Context.MODE_PRIVATE) ?: return
        with(sharedPref.edit()) {
            putInt(getString(R.string.activity_pref_key), 50)
            apply()
            super.onDestroy()
        }

    }
}

